src/storage/gpu_swap_history.cc

```diff
@@ -46,21 +80,66 @@ void MemHistory::PutRecord(handle_id_t handle_id, int device,
   record_idx[device]++;
 }
 
-// optimal algorithm: assume iterations remain the same; choose the handle
-// whose next reference is furthest in the future as victim.
-handle_id_t MemHistory::DecideVictim(std::unordered_set<handle_id_t> handles, int device) {
-  std::lock_guard<std::mutex> lock(mutex_[device]);
+handle_id_t MemHistory::LRU(std::unordered_set<handle_id_t> handles, int device) {
+  handle_id_t victim = -1;
+  while(lru_list[device].size() != 0 &&
+    handles.find(lru_list[device].back()) == handles.end()) {
+    handle_id_t temp_id = lru_list[device].back();
+    lru_map[device][temp_id] = lru_list[device].end();
+    lru_list[device].pop_back();
+  }
+  if(lru_list[device].size() == 0) {
+    std::cout << "No swappable handle found" << std::endl;
+  } else {
+    victim = lru_list[device].back();
+    lru_map[device][victim] = lru_list[device].end();
+    lru_list[device].pop_back();
+  }
+  return victim;
+}
+
+handle_id_t MemHistory::NaiveHistoryBased(
+  std::unordered_set<handle_id_t> handles, int device) {
   size_t latest_step = 0;
   handle_id_t latest_id = 0;
   for(auto &id : handles) {
-    MemHistory::MemRecord r = 
-        MemHistory::find(history[device][id], record_idx[device]);
-    if(r.record_step > latest_step) {
-      latest_step = r.record_step;
-      latest_id = r.handle_id;
+    MemHistory::MemRecord r = {0,MemHistory::GET_ADDR,0,record_idx[device],0};
+    std::vector<MemHistory::MemRecord>::iterator it =
+      std::upper_bound(history[device][id].begin(), history[device][id].end(),
+      r, CompareByStep);
+    if(it == history[device][id].end()){
+      if(record_idx[device] - history[device][id].back().record_step < 10) {
+        // Victim just used, skip
+        continue;
+      }
+      return id;
+    } else if(std::prev(it) != history[device][id].begin() &&
+        record_idx[device] - std::prev(it)->record_step < 10){
+      continue;
+    }
+    if(it->record_step > latest_step) {
+      latest_step = it->record_step;
+      latest_id = id;
     }
   }
   return latest_id;
+
+}
+
+// optimal algorithm: assume iterations remain the same; choose the handle
+// whose next reference is furthest in the future as victim.
+handle_id_t MemHistory::DecideVictim(std::unordered_set<handle_id_t> handles, int device) {
+  std::lock_guard<std::mutex> lock(mutex_[device]);
```

Comment:
I think it is not a good idea to always compare the environment variable (though it may not be costly). You can try function pointer. So you only need to compare once in the constructor and directly use the function pointer afterward. See https://stackoverflow.com/questions/2402579/function-pointer-to-member-function to know how to use function pointer for a class member.